MIPs/mip-mc-handshake.md

```diff
@@ -0,0 +1,255 @@
+---
+MIP: X
+Title: Wallet Authorization for a Multichain-Optimized MetaMask
+Status: Draft
+Stability: n/a
+discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions
+Author(s): Vandan Parikh (@vandan), Alex Donesky (@adonesky1)
+Type: Maintainer
+Created: 2024-01-25
+---
+
+## Summary
+
+A proposal to introduce an unrestricted `provider_authorize` method for applications to connect to MetaMask per ChainAgnostic Standards Alliance’s CAIP-25 standard. The `provider_authorize` method establishes an interface for permissioning application access to wallet-managed resources including accounts, RPC providers, and specific JSON-RPC methods across multiple chains (EVM and beyond) in a concurrent session. An application can request access to these resources by specifying scopes as defined in CAIP-217.
+
+This method is intended to be part of a new version of the MetaMask Wallet API that is optimized for multichain interactions.
+
+This proposal aims to implement CAIP-25's JSON-RPC Provider Authorization into MetaMask. For EVM (eip155) networks, the `rpcEndpoints` parameter is adapted to be compliant with the object structure defined in EIP-3085. Designing the interface in this way will provide a standardized approach for dApps to request access to networks and JSON-RPC methods.
```

Comment:
> the `rpcEndpoints` parameter is adapted to be compliant with the object structure defined in EIP-3085.

So, the whole EIP-3085 object instead of just `rpcURLs` array of strings? Maybe it might make for easier interop/conformance to just use the `rpcURLs` array for the CAIP-25 array of strings and stick the rest of the object (with or without the redundant member `rpcURLs`) elsewhere in the protocol? Apologies, I am lacking lots of context here.

Reply:
You could, for example, put a map in `sessionProperties` keyed to each rpcURL like so:

```js
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "provider_authorize",
  "params": {
    "optionalScopes":{ 
        "eip155:1":{
            ...
            "rpcEndpoints":[
                "https://rpc.linea.build"
            ]
            ...
        },
        "eip155:59144":{
            ...
            "rpcEndpoints":[
                "https://rpc.linea.build",
                "https://backup.infura.biz"
            ],
        },
        ...
    },
    "sessionProperties": {
        "rpcEndpointMetadata": {
            "https://rpc.linea.build": {
                "chainName":"Ethereum (Infura)",
                "rpcUrls":["https://rpc.linea.build"], /* keep or drop, i have no idea what's more ergonomic/secure/less-annoying */
                "nativeCurrency":{
                    "name":"ETH"
                    "symbol":"ETH"
                    "decimals":18
                },
                "iconURLs":["https://example.com/ethereum.svg"] /* (Optional) using an array to allow */
            },
            "https://backup.infura.biz": {
                ...
            }
        }
    }         
  }
}
```

Reply:
So per conversations with @hmalik88 we were thinking of approaching this in one of two ways:
1. Proposing to modify CAIP-25 such that the `rpcEndpoints ` property defined in [CAIP-217](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-217.md) would actually be an array of objects that would [conform to the EIP-3085 parameters](https://eips.ethereum.org/EIPS/eip-3085#wallet_addethereumchain-parameters)
2. Create a new CAIP where we specify a modification of the `rpcEndpoints` in the CAIP-25 scope object per what's described in  which could then be added to the 'sessionProperties' - something like `caipXYZCompliant = "true"` (totally unsure of how we'd want to play that) and then consumers can read `rpcEndpoints` according to the spec described in CAIP-XYZ

Reply:
Hey Alex! I can see CAIP-217 changing a lot or a little, nbd.  And yeah, if structured, security-sensitive things are going into sessionProperties, they should probably be specified in detail in a new CAIP, or even just the eip155/caip217.md profile in the namespaces repo if it's less a security-critical spec-spec and more just an explanation/semantic guidance.  i'm not wed to the exact design sketched above, it's just a sort of pattern or starting point i think might work for this particular kind of embellishment.  happy to schedule a meeting to discuss this-- take advantage of my being on a US time zone for a while!

Reply:
The people thinking about how to getting new kinds of information into CAIP-25 messages (whether in `sessionProperties` or otherwise might find [this exchange](https://github.com/ethereum/EIPs/pull/8387#discussion_r1551510667) interesting over on an EIP PR I opened yesterday